### ISSUE TITLE
Option to tolerate single lines with padded blocks

### ISSUE THREAD

**Issue by alphanull** [eslint/eslint#7145]
**What rule do you want to change?**

"padded-blocks"

**Does this change cause the rule to produce more or fewer warnings?**

fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**

New option. Could be similar to `{ "allExcept": ["singleLine"] }` in JSCS (http://jscs.info/rule/requirePaddingNewLinesAfterBlocks)

**Please provide some example code that this change will affect:**

`if (cond) { doSomething(); }`

**What does the rule currently do for this code?**

Warning / Error: "Block must be padded by blank lines."

**What will the rule do after it's changed?**

Emits no warning if using `{ "allExcept": ["singleLine"] }` option

**Comment by eslintbot** [comment]
Hi @alphanull, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by btmills** [comment]
Since the link seems to be broken at the moment, here's our [rule change proposal template](https://github.com/eslint/eslint/blob/master/templates/rule-change-proposal.md).

**Comment by alphanull** [comment]
No problem, I have updated my initial post. Hope its OK now.

**Comment by btmills** [comment]
Thanks @alphanull. Since this is similar to a JSCS option, I've gone ahead and added it to the JSCS milestone.

JSCS has five possible exceptions for the [requirePaddingNewLinesAfterBlocks](http://jscs.info/rule/requirePaddingNewLinesAfterBlocks) rule:

> - "inCallExpressions" Blocks don't need a line of padding in function argument lists
> - "inNewExpressions" Blocks don't need a line of padding in constructor argument lists
> - "inArrayExpressions" Blocks don't need a line of padding in arrays
> - "inProperties" Blocks don't need a line of padding as object properties
> - "singleLine" Blocks don't need a line of padding if they are on a single line

Is this something where we should add them all at once, or just as requested (`"singleLine"`) in this case?

**Comment by alphanull** [comment]
`"singleLine"` would be what bothers me the most right now. But in this case other options could also be useful to other folks (and maybe even to me, not sure ATM), so why not add them all at once? Then it would also be fully compatible to JSCS. In addition, I would also like to see an option to specify minumum number of lines within a block, like described in #7146. And yes, adding options to enforce padding _after_ and/or _before_ blocks would also be nice (#6445). So it might look like `padded-blocks`  is ready for an overhaul ;)

**Comment by platinumazure** [comment]
We just need to be careful about where we might be covering the other JSCS options in other ESLint rules. For example, "inProperties" might be covered by ESLint rule `object-curly-newline`.

**Comment by kaicataldo** [comment]
I wonder if, to get the ball rolling, we could just implement `singleLine` (which seems like the one that is most useful) now and wait and see if there are requests for the other options in the future. Since they're all behind their own unique flags and off by default, they can all be implemented independently and none are breaking changes.

I can champion this change if the above sounds good!

**Comment by alphanull** [comment]
Yeah, that would be great! I agree with `singleLine` being the most imporant thing to implement here.

**Comment by vitorbal** [comment]
@kaicataldo I think that's a good idea! 👍 from me.

**Comment by not-an-aardvark** [comment]
This seems like a good idea, but I don't think `allExcept: ["singleLine"]` is a good API for it. An array containing a single string seems like an inelegant API to me; it would be better to have explicit options instead.

How about something like this instead?

```yml
rules:
  padded-blocks: [error, always, {allowSingleLineBlocks: true}]
```

**Comment by kaicataldo** [comment]
@not-an-aardvark Agreed. My proposal is to implement the feature, but wasn't necessarily intending the same API. Sorry, should have made that clearer!

**Comment by kaicataldo** [comment]
Do we want to continue pursuing this now or wait until we see how [this discussion](https://github.com/eslint/eslint/issues/7356) resolves?

**Comment by sscaff1** [comment]
Almost the same as #7775

**Comment by platinumazure** [comment]
Ah crap, I didn't do my due diligence correctly.

@sscaff1 @kaicataldo How should we handle this? Can this issue (or #7775) move forward by itself, or do we really need to tie this to #7356?

**Comment by sscaff1** [comment]
Either issue will solve it for me so it doesn't matter to me. I think either my issue or @alphanull issue can move forward by itself. My issue probably introduces a more flexible API although I'm unclear of what the use case is to require 3 lines for example before padding (although I'm sure it's out there somewhere).

**Comment by qlaire** [comment]
I'm going to work on this issue.

**Comment by ilyavolodin** [comment]
I know this has been accepted a while back, but I have a new proposal for this. How about deprecating existing `blocks`, `classes` and `switches` options, and adding `exceptions` option of type Array, that accepts `ASTNode|"single"|"block"`? That should cover all JSCS cases, and bring this rule inline with other rules we have.

**Comment by kaicataldo** [comment]
@ilyavolodin I think this makes sense, though I'm a little bit concerned that we've made a few changes that require the user to look up the ESTree node type to configure ([example](https://github.com/eslint/eslint/issues/7356)). I wonder if we should create some docs surrounding ASTs to make this more easily discoverable.

**Comment by ilyavolodin** [comment]
@kaicataldo Sure, that would be fine. We have a bunch of rules that already accept AST node types as parameters, so I think that would be a good practice to follow.

**Comment by kaicataldo** [comment]
@ilyavolodin Do you mind fleshing out your proposal? It sounds good to me, but I think it would be helpful for contributors to have a clearer picture of what's being asked for here. 

@eslint/eslint-team Thoughts on this?

**Comment by kaicataldo** [comment]
@ilyavolodin I think the thing we need to hash out here is what an `exception` actually means. The current rule actually has three states when the object form is specified - padded, unpadded, and ignored (when not specified).

Were you thinking that `exception`s would be ignored or would enforce the opposite of the first string configuration option?

**Comment by kaicataldo** [comment]
@eslint/eslint-team I'd really like to get this implemented. Given the complexities in my previous comment, I wonder if it would be beneficial to implement either the proposal in https://github.com/eslint/eslint/issues/7145#issuecomment-258700883 or to solve this with the proposal in https://github.com/eslint/eslint/issues/7775?

**Comment by ilyavolodin** [comment]
Really sorry, I totally lost track of this issue. My proposal was to change the options to `always|never` as a first argument, and an array of `ASTNode|"single"|"block"` as a second option. Anything in the array would be the opposite of the first value.

**Comment by nzakas** [comment]
@kaicataldo @ilyavolodin are either of you still interested in implementing this?

**Comment by kaicataldo** [comment]
Apologies for dropping the ball as the champion of this issue. Given that this is a JSCS compatibility issue, I would like to see this through. 

While I appreciate @ilyavolodin's solution, I would like to propose we go with the previous solution as  described [here](https://github.com/eslint/eslint/issues/7145#issuecomment-258700883). We can always change the rule in the future if need be, and this would buy us JSCS compat without having to make major changes to the implementation.

**Comment by richie3366** [comment]
I wanted that option, so I implemented it myself (see PR #11243).

The tests & docs are not shipped in my PR, so it will need some contribution from you guys.

**Comment by kaicataldo** [comment]
I'm not entirely sure what to do here. There was initially consensus, and then a new approach was suggested, but it looks like the conversation has stalled. @ilyavolodin Any other thoughts on my questions above?

**Comment by richie3366** [comment]
@ilyavolodin And what about @not-an-aardvark proposal?

It introduces less breaking changes and it's already PR'd and ready to review (and merge if everyone agrees on this proposal & on the PR review).

Sorry if I may have done a wrong move by making a PR on a proposal which hadn't been agreed upon yet.

**Comment by ilyavolodin** [comment]
Ah, thanks. I forgot about that. Yes, I would prefer syntax suggested by @not-an-aardvark It's more inline with other rules we have.

**Comment by kaicataldo** [comment]
Sounds like we've come to a consensus and have accepted the implementation outlined [here](https://github.com/eslint/eslint/issues/7145#issuecomment-258700883).
